expression/builtin_compare.go

```diff
@@ -2429,7 +2431,20 @@ func CompareString(sctx sessionctx.Context, lhsArg, rhsArg Expression, lhsRow, r
 	if isNull0 || isNull1 {
 		return compareNull(isNull0, isNull1), true, nil
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	ret := compareStringWithTrim(arg0, arg1, lhsArg.GetType(), rhsArg.GetType())
+	return int64(ret), false, nil
+}
+
+// According to https://dev.mysql.com/doc/refman/8.0/en/char.html,
+// we need to trim trialing space when compare non-binary strings.
+func compareStringWithTrim(lhs string, rhs string, lhsFt *types.FieldType, rhsFt *types.FieldType) int {
```

Comment:
Seems we should consider trimming for both side together, I'm not sure the detailed logic of MySQL, but here is a case I found:

```
(MySQL 8.0)
mysql> select "a " = binary("a ");
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.01 sec)

mysql> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

(TiDB with this PR)
tidb> select "a " = binary("a ");
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   0 |
+---------------------+
1 row in set (0.00 sec)

tidb> select "a" = binary("a");
+-------------------+
| "a" = binary("a") |
+-------------------+
|                 1 |
+-------------------+
1 row in set (0.00 sec)

tidb> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

(TiDB without this PR)
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.01 sec)

tidb> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

tidb> select "a " = binary("a");
+--------------------+
| "a " = binary("a") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.01 sec)

tidb> select "a" = binary("a");
+-------------------+
| "a" = binary("a") |
+-------------------+
|                 1 |
+-------------------+
1 row in set (0.00 sec)
```